sdk/python/kfp/_local_client.py

```diff
@@ -0,0 +1,503 @@
+
+        return group_dag
+
+    def _create_op_dag(self, p: dsl.Pipeline) -> _Dag:
+        """ Create the DAG of the pipeline ops. """
+        dag = _Dag(p.ops.keys())
+
+        for op in p.ops.values():
+            # dependencies defined by inputs
+            for input_value in op.inputs:
+                if isinstance(input_value, dsl.PipelineParam):
+                    input_param = self._extract_pipeline_param(input_value.pattern)
+                    if input_param.op_name:
+                        dag.add_edge(input_param.op_name, op.name)
+                    else:
+                        logging.debug("%s depend on pipeline param", op.name)
+
+            # explicit dependencies of current op
+            for dependent in op.dependent_names:
+                dag.add_edge(dependent, op.name)
+        return dag
+
+    def _alter_output_file_path(
+        self, run_name: str, op_name: str, output_file: str
+    ) -> str:
+        """Alter the file path of output artifact to make sure it's unique in local runner.
+
+        kfp compiler will bound a tmp file for each component output, which is unique
+        in kfp runtime, but not unique in local runner. We alter the file path of the
+        name of current run and op, to make it unique in local runner.
+        """
+        return re.sub(
+            "/tmp",
+            "/{artifact_root}/{run_name}/{op_name}".format(
+                artifact_root=self._artifact_root,
+                run_name=run_name,
+                op_name=op_name.lower(),
+            ),
+            output_file,
+        )
+
+    def get_output_file_path(
+        self,
+        run_name: str,
+        pipeline: dsl.Pipeline,
+        op_name: str,
+        output_name: str = None,
+    ) -> str:
+        """ Get the file path of component output. """
+
+        op_dependency = pipeline.ops[op_name]
+        if output_name is None and len(op_dependency.file_outputs) == 1:
+            output_name = next(iter(op_dependency.file_outputs.keys()))
+        output_file = op_dependency.file_outputs[output_name]
+        altered_output_file = self._alter_output_file_path(
+            run_name, op_name, output_file
+        )
+        return altered_output_file
+
+    def _op_cmd_locally(
+        self,
+        run_name: str,
+        pipeline: dsl.Pipeline,
+        op: dsl.ContainerOp,
+        stack: Dict[str, Any],
+    ) -> List[str]:
+        """ Generate shell command to run the op locally. """
+        cmd = op.command + op.arguments
+
+        # In debug mode, for `python -c cmd` format command, pydev will insert code before
+        # `cmd`, but there is no newline at the end of the inserted code, which will cause
+        # syntax error, so we add newline before `cmd`.
+        for i in range(len(cmd)):
+            if cmd[i] == "-c":
+                cmd[i + 1] = "\n" + cmd[i + 1]
+
+        _arg_pairs = self._get_keyword_arguments(cmd)
```

Comment:
Let's name these variables arg_pairs and fixed_arg_value instead, and reserve the `_` prefix for private members and methods.